### PR TITLE
Non-Blocking IO and Task Scheduling Enhancements ... and anything else that fits in the kitchen sink.

### PR DESCRIPTION
**Pull request by directorm**
THIS TIME IT WORKS :)
<sup>(re: a few days ago and a botched commit)<sup>

Complete overhaul of Networking, Reporting, and Threading.

**Reduced running threads by almost 100 threads.**

1. Made Non-Blocking incoming TCP replicator.

The incoming TCP is done with NIO socket channels ensure that requests are handled as the packets arrive, without blocking, asynchronously on the system fork-join pool.

The outgoing TCP is now push driven by Striped Executor (SE) tasks and properly connects/reconnects immediately handling typical errors. This is higher throughput with minimal blocking and no dedicated threads necessary.

2. Revised outgoing TCP and UDP for better control flow and error handling/recovery.

3. Moved a lot of tasks that had dedicated threads into scheduled tasks.

All tasks now driven through SE  - in the future it can become a dynamic load balancer responding to various operational loads.

4. Revised Node.java to submit request/reply/processing/broadcast/tips jobs to SE for enhanced queue management.

5. Created Non-blocking DNS Refresher - no more 30 second DNS refresh waits.

6. Separated out Neighbor management functions into a dedicated class - NeighborManager.java.

7. Made small changes wherever code was unclear.

8. Merged fixed Hash class and corrupted packets as per recent github issues.

9. Introduced some locking in critical classes and cleaned up encapsulation.

10. Created functional interfaces and some useful utility classes.

11. It now accepts a proper command line according to the spec - just an INI file needs to be specified and the reporting for bad neighbor configs and INI & PORT detection has been improved.

12. A lot more live reporting of IO and task load balance.

Logging is a bit verbose - but left it in to show the new inner workings. You will need to go full screen!

New package dirs: exec, scheduledReports, scheduledTasks, funcinterfaces, textutils - are all new. All old networking code cleaned out.

# How Has This Been Tested?
I have tested it on a private testnet with multiple instances connecting with each other.

# Checklist:
- [X ] My code follows the style guidelines of this project - mostly.
- [X ] I have performed a self-review of my own code
- [ X] I have commented my code, particularly in hard-to-understand areas
- [X ] New and existing unit tests pass locally with my changes

I don't expect that this will get adopted wholesale -- this is for discussion and as a demo to inspire future changes. 

There are too many revisions here all over the project (  ... I know!) It was a NIO job and it turned into a very large one shot revision. It could use more input and someone to clean it up with a different set of eyes.

### CODE REVIEW

**Comment by GalRogozinski** [comment]
Thanks!

Since this a big PR it will take some time until we get to it. As you know, we will have to compare it with th0br0#1. In the meanwhile any work by the community to compare and test the two implementations will be appreciated!

File: src/main/java/com/iota/iri/network/NeighborManager.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+
+    public boolean add(Neighbor neighbor) {
+        return neighbors.add(neighbor);
+    }
+
+    public boolean removeNeighbor(final URI uri, boolean isConfigured) {
+        final Neighbor neighbor = newNeighbor(uri, isConfigured);
+        if (uri.getScheme().equals("tcp")) {
```
Comment (akashgoswami) [inline src/main/java/com/iota/iri/network/NeighborManager.java]: what happens in case of UDP neighbors?

**Comment by akashgoswami** [comment]
Starting with default argument throws an exception
The default SEND_LIMIT is set to "-1.0"


java -Xms256m -Xmx2048m  -jar iri-1.4.2.2.jar -p 14625
06/21 18:23:06.074 [main] INFO  com.iota.iri.IRIImpl - Port setting obtained from commandline: 14625
06/21 18:23:06.078 [main] INFO  com.iota.iri.IRIImpl - Welcome to IRIImpl 1.4.2.2
06/21 18:23:18.766 [main] ERROR com.iota.iri.IRIImpl - Initialization exception: TPS must be a positive non-zero number: -75
java.lang.IllegalArgumentException: TPS must be a positive non-zero number: -75
        at com.iota.iri.network.exec.SendTPSLimiter.<init>(SendTPSLimiter.java:27) ~[iri-1.4.2.2.jar:na]
        at com.iota.iri.network.exec.SendTPSLimiter.<init>(SendTPSLimiter.java:22) ~[iri-1.4.2.2.jar:na]
        at com.iota.iri.network.Node.<init>(Node.java:129) ~[iri-1.4.2.2.jar:na]
        at com.iota.iri.Iota.<init>(Iota.java:79) ~[iri-1.4.2.2.jar:na]
        at com.iota.iri.IRIImpl.init(IRIImpl.java:85) ~[iri-1.4.2.2.jar:na]
        at com.iota.iri.IRIImpl.main(IRIImpl.java:46) ~[iri-1.4.2.2.jar:na]

**Comment by GalRogozinski** [comment]
This was done in #1072

**Pull request closed without merging**